[Bongf]미션 3 : 체스판초기화

bong6981 opened the pull request:
안녕하세요. 백엔드반 Bongf입니다. 부족한 코드 피드백 주셔서 감사하다는 말씀 드립니다. 
코드를 계속해서 수정하다 보니 커밋이 너무 많고, 두서 없습니다. 
다음 번엔 수정도 관련있는 것들끼리 묶어 커밋하는 방식으로 시도해보겠습니다. ( 코드 처음에 작성할 때 더 깔끔하게 작성해야겠다는 생각이 들었습니다. )

고민 되었던 부분은 색의 representation을 어디에 작성할 것 인가 이었습니다. 
- 우선은 색에 따라 respresentation이 나뉘게 되고, 아직은 다른 말이 없어 representation이 p/P 밖에 없어 Color 클래스에서 선언해도 될 것 같아 Color에 작성하였습니다. 
- 접근제어자는 최대한 좁게 설정하여 메소드의 접근제어자도 private으로 설정하였다가, test를 위해서 default로 바꿨습니다. 
    - 관련 내용을 학습했을 때는 왜 private 메소드가 되어야 하는지 생각해 보라는 답을 얻어 이에 대해 고민하게 되었습니다.   
    - 이 메소드들이 game을 시작하면 계속 사용될 것이라 생각되어 public으로 설정했고, 다만 클래스 내부에서 사용되는 메소드들에 대해서만 private으로 설정했습니다.

bong6981 committed:
Edit code for conflicts1

mission2의 풀리퀘가 수정되기 전에 다른 내용을 수정하여 merge할 때
conflict를 일으키기 위한 코드 변경

bong6981 committed:
Pawn 색에 따른 출련 문자 부여

- PawnTest에서 getRepresentation과 해당 색의 representations 값을
  비교하는 test 작성
- Pawn에서 getRepresentation 메서드 추가
- Pawn 클래스안에 final char WHITE_REPRESENTATION = 'p'와
  BLACKREPRESENTATION = 'P'를 각각 선언
- Pawn의 멤버변수 representation 추가 ( 각 pawn의 색 대표자(p혹은 P)를
  가질 변수)
- 생성자에 pawn의 색에 따라 각pawn의 respresentation가 설정될 수 있도록
  구현

bong6981 committed:
색의 representations을 enum Color에 포함시킴

- 해당 색과 그것의 대표자가 더 연관성 있다고 판단되엇고
- 색을 나타내는 char가 PawnTest 안에 선언될 필요를 못 느껴서
- enum Color의 멤버변수에 representation추가
- 생성자에서 representation을 설정하게 해줘서 각각의 객체(WHITE, BALCK)에 representation추가
- Pawn의 말의 색을 나타내는 colorRepresentation chr 변수추가
- Pawn에 pawn 생성시 pawn의 색에 따라 해당 색의 representations을 갖도로
  구현

bong6981 committed:
체스판 초기화

- BoardTest에 initialize() test 코드를 작성해서 board를 초기화 시키면
  8개의 하얀 폰과 8개의 검은 폰이 추가 되어
  getWhite/BlackPawnsResult() 실행시 하얀폰은 'p'로 검은폰은 'P'로
  repsresentation 값을 얻는지 확인한다.
- Board에 initialize()메소드 구현
	- initialize()실행시 하얀폰과 검은폰 각각 8개의 폰은 생성하고
	  addWhite/BlackPawn으로 이를 보드의 리스트에 각각
	  추가시킴(whitePawns와 blackPawns로 별도의 리스트로 구분)
- 기존의 테스트코드와 구현코드가 오류가 나서 현상황에 맞게 수정해줌
	- size()는 whitepawns와 blackpawns의 사이즈를 더함
	- findPawn도 findWhitePawn과 findBlackPawn으로 구분하여 각각의
	  리스트에서 인덱스로 각각의 폰을 찾게함
- getWhitePawnsResult()d와 getBlackPawnsResult()로 각 색별 리스트에 있는
  pawn을 StringBuufer에 담아서 toString으로 변환

bong6981 committed:
Board 클래스 중복 리팩토링

- getBlack/WhitePawnsResult()에서 중복이 일어났고, 이 메서드는 1)
StringBuffer 생성 2) black/whitePawns 리스트의 각 pawn들의
representations 값을 얻어와서 1-2) StringBuffer에 넣어주는 일을 하고
있었다.
- 크게보면 두 가지 일을 하고 있다고 생각하여 메소드를 나눠줬다.
- toRepresentations를 추가하여 pawnList를 매개변수로 받아, pawnList에서
  각 pawn에 대한 representation 값을 얻어 char[]에 넣어 반환하도록 했다.
- 결과적으로 getBlack/WhitePawnsResult에는 StringBuffer를 생성하고
  그곳에 toRepresentations에서 반환받은 char[]의 값을 넣어주는 것으로
  변경했다.

bong6981 committed:
print() 구현

- emptyBoard로 보드를 .으로 채워 넣고
- 각각의 행위 위치에 getBlackPawnsResult와 getWhitePawnsResult()를
  넣는다

bong6981 committed:
main() 구현

- main()메소드를 작성하여 새로운 클래스를 추가하고 시작, 종료 기능 제공
- while문으로 반복하다 사용자가 "end" 입력시 break되도록 구현

bong6981 committed:
geWhite/BlackPawnsResult()리팩토링

- 두 메소드에 겹치는 부분이 있다고 판단되어 getPawnsResult() 메서드로
  공통 부분을 빼줌

bong6981 committed:
Test 이름이 작성되지 않아 Test이름 작성

Test 이름이 미작성되어 내용에 맞게 Test 이름 작

bong6981 committed:
addBlack/WhitePawn이 코드 수정

addBlack/WhitePawn()이 각자 색이 맞을 때만 확인하도록 코드 수정

bong6981 committed:
print()메소드와 매직넘버 수정

- 다른 분들의 피드백을 보고 print()와 매직넘버 수정
- print()를 for문으로 돌려서 빈 체스판을 올리고 말을 올리는 식으로
  만들었었는데, 그렇게 하다보니 매직넘버가 많아져서 수정
- 직접 print("......")해주는 것이 더 직관적으로 이해하기 쉽다는 것을
  알게되었다.
- board의 열에 해당되었던 8이 매직넘버이기 때문에 이를 BOARD_FILE
  로선언해줌

bong6981 committed:
Pawn의 getColorRepresentation을 getRepresentation으로 변경

- 체스 규칙상 색깔을 나타내는 것 외에 구분자를 사용할 일이 없을 것 같아
  getRepresentation으로 이름 변경

bong6981 committed:
initialize() 메소드 수정

- for문이 중복되어 addWhit/BlackPawn() 아규먼트로 Pawn생성자를 넣어줌

bong6981 committed:
코드 수정

- representations()메소드를 getRepresentations()으로 수정
	- 원래는 for(representation : representations) 가독성을 위해
	  메소드 명을 representations으로 설정
	- 하지만, 메소드의 기능상 representations를 얻는 기능이므로,
	  getRpresentations으로 수정
- test 클래스는 일반적으로 디폴트로 설정해야 한다는 것을 알게 되어
  디폴트로 접근제어자 변경
- 테스트코드가 fail되어도 그 전부를 실행에 어디서 틀려줄지 알려주기
  때문에 assertAll을 사용하여 여러 assertions을 묶어줌
- 그밖에 불필요한 개행 삭제

bong6981 committed:
리팩토링 DisplayName, 메소드이름, assertAll

- displayname(흰색 pawn 생성)이기능(검은색, 흰색 pawn 생성)을 포함하지
  못하여 기능을 포함하도록 이름 수정
- 지난번에 지적받았던 test 메소드 이름이 적합하지 못하다고 판단되어 더
  간결한 방식으로 표현 (displayName에서 자세한 내용이 설명되고 있다고
  생각함)
- 도중 하나가 fail되어도 끝까지 체크하여 어떤 것이 문제인지 알려주는
  assertAll을 사용함

bong6981 committed:
assertAll로 코드 수정

- 미처 수정하지 못한 assertThat을 assertAll로 묶어줌
- assertAll은 도중에 assertions이 fail해도 끝까지 검사하고 어디에서
  fail했는지 정확히 알 수 있다.

LarryJung commented on src/main/java/chess/Board.java:
```diff
@@ -1,22 +1,89 @@
+        return representations;
+    }
+
+    public void addWhitePawn(Pawn pawn) {
+        if (pawn.getColor() == Color.WHITE) {
+            whitePawns.add(pawn);
+        }
+    }
+
+    public void addBlackPawn(Pawn pawn) {
+        if (pawn.getColor() == Color.BLACK) {
+            blackPawns.add(pawn);
+        }
+    }
```
if 문으로 체크하는 부분은 잘 하셨습니다! ㅎㅎ 
한마디 덧붙여보자면.. 색이 다른 경우는 exception을 발생한다던지? 아니면 올바른 add를 했다는 의미로 true, false 등을 리턴하는 식으로 정보를 추가해주어도 좋을 것 같습니다. (add 실행 후 결과처리할 때도 확실하고, 테스트할때도 써먹을 수 있고용)

bong6981 replied:
말씀주신 대로 예외를 발생시켰고, test도 해보았습니다.!

LarryJung commented on src/main/java/chess/Board.java:
```diff
@@ -1,22 +1,89 @@
+        }
+    }
+
+    public void print() {
+        StringBuilder result = new StringBuilder();
+        String emptyLine = "........";
+        result.append(emptyLine)
+                .append("\n")
+                .append(getWhitePawnsResult())
+                .append("\n")
+                .append(emptyLine)
+                .append("\n")
+                .append(emptyLine)
+                .append("\n")
+                .append(emptyLine)
+                .append("\n")
+                .append(emptyLine)
+                .append("\n")
+                .append(getBlackPawnsResult())
+                .append("\n")
+                .append(emptyLine);
+        System.out.println(result);
```
board의 print()는 String 만 리턴하고, System.out.println을 사용하는 출력기능은 ChessGame쪽에서 담당하게 하면 어떨까요?

현재 요구사항에서는 충분히 괜찮지만.. 나중에 게임화면 출력용 ui앱을 만든다고 하면.. 게임로직과 입출력은 한 클래스에 있는 것보다는 분리하는게 나을것 같네요! (당장 안하셔도 되고ㅎㅎ 나중에 복잡성이 늘어나면 입출력 로직을 떼어낼 수도 있겠다라는 생각 가지고 계시면 좋을것 같네용)

bong6981 replied:
오! 맞는 말씀입니다.  다음단계에서 반영해보도록 하겠습니다. 지금은 우선 밑에 코멘트  test 코드에서 print를 하도록 바꾸었습니다. 그리고 수정과정에서 조건을 읽다보니 string으로 반환하고 이를 test 코드에서 출력한다는 부분이 있어 현재는 print() 메소드에는 System.out.println부분을 제거했습니다.  게임화면 출력 ui를 만들 때에는 반드시 분리해야겠다는 생각이들었습니다. 감사합니다.

LarryJung commented on src/main/java/pieces/Color.java:
```diff
@@ -1,5 +1,16 @@
 
 public enum Color {
-    WHITE, BLACK
+    WHITE('p'), BLACK('P');
+
+    private char representation;
+
+    Color(char representation) {
+        this.representation = representation;
+    }
+
+    public char getRepresentation() {
+        return this.representation;
+    }
+
```
현재는 이렇게 가셔도 깔끔하고 좋습니다 ㅎㅎ

LarryJung commented on src/main/java/pieces/Pawn.java:
```diff
@@ -2,16 +2,23 @@
 
 public class Pawn {
     private final Color color;
+    private final char Representation;
```
representation은 불필요한 멤버 변수입니다. (변수이름이 대문자..!)

color 객체가 representation 값을 가지고 있으니 꼭 pawn이 가지고 있어야 할 필요는 없어 보이네요.
오히려 pawn의 color가 바뀌었을때 representation을 재할당 해줘야 하는점에서 오류가 생길 수 있기도 하고요~

bong6981 replied:
헉. ColorRepresentation 에서 수정을 하다 발생한 오류 같습니다. 조금 더 꼼꼼하게 작성하겠습니다.

representation을 어디에 둘 것인지에 대한 고민의 잘못된 결과로 이렇게 될 것 같습니다.
말씀해 주신 대로 생각해보니 
- Color 에 따라서 representation을 갖고 있으니
-  Pawn이 Color 값을 가지면 그에 따라 자동으로 representation을 갖게 된 것이네요 
재할당해줘야 하는 이유가 전혀 없었네요 
코드 실행 결과에만 급급하다보니 나오게된 결과 같습니다.
앞으로는 로직을 좀 더 먼저 생각해보아야 할 것 같습니다.!

LarryJung commented on src/main/java/pieces/Pawn.java:
```diff
@@ -2,16 +2,23 @@
         return color;
     }
+
+    public char getRepresentation() {
+        return Representation;
+    }
```
`return this.color.getRepresentation()` 이면 충분하겠죠?

bong6981 replied:
위에 주신 코멘트와 함게 수정했습니다~!

LarryJung commented on src/test/java/chess/BoardTest.java:
```diff
@@ -10,28 +10,46 @@
+        );
+    }
+
+    @Test
+    @DisplayName("board를 초기화 하고 print한다")
+    void print() {
+        board.initialize();
+        board.print();
```
board.print() 를 제대로 테스트하고 있지 못한 것 같네요~ 테스트코드는 print()가 어떻게 나오든 assert로 검사하는게 없으니 무조건 통과가 될 것 같은데요~ 이럴 때 board.print()가 출력물 String을 return 한다면? 테스트를 할 수 있을 것 같은데..

public void 메서드를 작성할 때 이것을 어떻게 테스트 할 수 있을지 염두하시는 것은 매우 좋은 습관입니다!

bong6981 replied:
어떻게 test할지 더 고민해보겠습니다! 사실 처음에는 String을 return하겠다고 짰다가 가이드 코드가 위와 같이 되어 있다고 해서 이런 의도였군! 하면서 수정했습니다.  하지만 조건을 꼼꼼히 읽어보니, '테스트 코드에서 반환한 값을 System.out.println으로 출력한다.' 부분을 확인할 수 있었습니다. 앞으로는 조건을 더 꼼꼼히 읽겠습니다. 코멘트 덕분에 print 메서드는 어떻게 테스트하는지도 조금이지만 공부할 수 있었습니다. 감사합니다 : )

LarryJung replied:
넵 프린트 뿐 아니라 앞으로 로직 짜실 때 값을 반환하지 않는 void메서드를 자주 작성하실 텐데 그때 꼭 void가 필요한지? 구조를 다르게 해서 return 을 하게하면 어떨지? 많이 고민해보세요!

bong6981 replied:
네 void 메서드 작성할 때마다 고민해보겠습니다!

LarryJung commented on src/test/java/pieces/PawnTest.java:
```diff
@@ -3,25 +3,31 @@
-        verifyPawn(Color.BLACK);
+        verifyPawn(Color.WHITE, Color.WHITE.getRepresentation());
+        verifyPawn(Color.BLACK, Color.BLACK.getRepresentation());
     }
 
-    void verifyPawn(Color color) {
+    void verifyPawn(Color color, final char colorRepresentation) {
         Pawn pawn = new Pawn(color);
-        assertThat(pawn.getColor()).isEqualTo(color);
+        assertAll(
+                () -> assertThat(pawn.getColor()).isEqualTo(color),
+                () -> assertThat(pawn.getRepresentation()).isEqualTo(colorRepresentation)
+        );
```
color 객체가 가진 정보가 있는데 verifyPawn에 representation 인자가 추가된 점이 아쉽습니다! 
OOP 하실때 객체가 가진 정보(멤버 변수)를 스스로 잘 활용하게(메서드) 하도록 신경써 주세요~! 
다소 추상적인 이야기지만 인사이트가 되었으면 좋겠습니다. ㅎㅎ

```java
verifyPawn(Color.WHITE);
verifyPawn(Color.BLACK);

void verifyPawn(Color color) {
    Pawn pawn = new Pawn(color);
    assertAll(
                () -> assertThat(pawn.getColor()).isEqualTo(color),
                () -> assertThat(pawn.getRepresentation()).isEqualTo(color.getRepresentation())
        );
}
```

bong6981 replied:
Color로 지정해 두었는데 이렇게 쓸 생각을 못했네요! 아직은 명확히 이해하지 못했지만, 'OOP 할 때 객체가 가진 정보를 스스로 잘 활용하게 하자.  test 코드에도 새로운 변수를 선언하기 보다는 객체들을 활용할 수 있도록 하자' 정도로 이해하고 적용해보겠습니다! 이해 안가는 부분도 좀 더 공부해보겠습니다 : )

LarryJung reviewed (changes requested):
수고 많으셨습니다~ 전체적으로 코드가 눈에 잘 보이고 깔끔하게 짜려고 노력하신게 느껴졌습니다.

다만 color 의 representaion을 활용하는 부분에서 조금 아쉬웠구요! 코멘트 드렸으니 참고하셔서 수정부탁드립니다ㅎㅎ

bong6981 committed:
피드백 반영하고 코드 수정

-1) addWhite/BlackPawn 에 잘못된 색 입력하였을 때, 예외 발생하도록 구현
-2) print()메서드에서 String을 return하게 하고, 이를 test코드에서 확인
	- 구현 조건에도 더 맞는 것이라는 것을 확인
-3) Representation 변수 삭제
	- Color에서 이미 값을 갖고 있기 때문에 색을 갖으면 자동으로
	  representation을 갖게된다는 것을 놓침.
-4) test코드에 직접 'p'/'P"를 매개변수로 전달해 representation을
확인하였던 것을 Color enum을 활용해서 전달

bong6981 commented:
말씀주신대로 수정사항 반영했습니다. 
틀린 부분을 정확히(따뜻한 온도로..) 짚어 주셔서 코드가 깔끔해지고 오류도 수정된 것 같습니다. 감사합니다.
말씀주신 부분에 대해 미흡한 점이 많아 앞으로도 더 학습해보도록 하겠습니다.

---

[수정사항]
1. addWhite/BlackPawn() 에 잘못된 색 입력하였을 때, 예외 발생하도록 구현
  - 예외를 발생시키지 않으면, 해당 메소드에 잘못된 값을 입력했다는 알림(?)이 없어 이를 추가하고 이 예외 발생 부분을 test 코드에서 확인함 

2. print()메서드에서 String을 return하게 하고, 이를 test코드에서 확인
- 미션 조건을 꼼꼼히 읽어보았을 때 구현 조건에도 더 맞는 것이라는 것을 확인
-게임로직과 입출력을 분리하는 조건에 대해서도 고심해야 함. 

3. Representation 변수 삭제 
- 오타로 대문자를 갖고 있던 변수 삭제 
-  해당 변수가 사실 enum Color에 구현되어 있어 필요없는 부분이었음에도 구현한 불필요한 부분이라 삭제 
- Color를 활용

4.  test코드에도 respresentation을 'p'/'P'로 하드코딩하지 않고 Color에 있는 값으로 활용함 
- '객체가 가진 정보(멤버 변수)를 스스로 잘 활용하게(메서드) 하도록 신경쓰기'에 대해서 학습하고 숙지해야 함

LarryJung commented on src/main/java/chess/Board.java:
```diff
@@ -48,15 +49,17 @@ private String getPawnsResult(List<Pawn> pawns) {
-        if (pawn.getColor() == Color.WHITE) {
-            whitePawns.add(pawn);
+        if (pawn.getColor() != Color.WHITE) {
+            throw new InvalidParameterException("pawn의 색을 확인해 주세요");
```
어떤 색깔이 넘어왔는지 메세지 안에 포함시켜줘도 좋습니다. ㅎㅎ

bong6981 replied:
다음 미션에서 해당 부분 수정하겠습니다~!

LarryJung commented on src/test/java/chess/BoardTest.java:
```diff
@@ -6,7 +6,10 @@
 import pieces.Color;
 import pieces.Pawn;
 
+import java.security.InvalidParameterException;
```
InvalidParameterException <- 가져다 쓰고 싶게 생긴 익셉션 이름이긴 한데요, 

독스 들어가보시면 자바 시큐리티 모듈 내 암호화 관련 로직에 자기네들이 쓰려고 만든 일종의 커스텀 익셉션입니다..ㅎㅎ

java.lang 에 들어있는 IllegalArgumentException을 사용하는것이 더 일반적입니다~

bong6981 replied:
오 둘 중에 하나를 고민했는데 학습이 부족했던 결과입니다. 더 학습하고 수정하겠습니다~!

LarryJung reviewed (approved):
수정사항 확인했습니다~ 수고하셨구요! 코멘트를 조금 드리긴 했는데 다음 단계 하셔도 좋을것 같습니다!^^
